### PR TITLE
Fix bug in Slider where it does not run in API < 16 and does not load in API < 8

### PR DESCRIPTION
**Pull request by halatmit**

### CODE REVIEW

**Comment by halatmit** [comment]
issue #463

File: appinventor/components/src/com/google/appinventor/components/runtime/Slider.java [hunk 1]
```diff
@@ -109,6 +130,25 @@ public Slider(ComponentContainer container) {
...
+
+    if (DEBUG) {
+      Log.d(LOG_TAG, "API level is " + SdkLevel.getLevel());
+    }
+
+
+    // Warn the user if the API level is less than Jellybean, since getThumb
+    // does not work before API level 16.  Note that this is *not* sufficient in dealing
+    // with very old phones.   See the SeekBarHelper class above.
+    if (SdkLevel.getLevel() < SdkLevel.LEVEL_JELLYBEAN) {
```
Comment (jisqyv) [inline appinventor/components/src/com/google/appinventor/components/runtime/Slider.java]: @halatmit I'm not sure if this is appropriate. I just tested this code on Gingerbreak (2.3.3) and I received the warning, but the slider worked as expected.

File: appinventor/components/src/com/google/appinventor/components/runtime/Slider.java [hunk 2]
```diff
@@ -4,6 +4,7 @@
 import android.graphics.PorterDuff;
 import android.graphics.drawable.ClipDrawable;
 import android.graphics.drawable.LayerDrawable;
+import android.media.SoundPool;
```
Comment (jisqyv) [inline appinventor/components/src/com/google/appinventor/components/runtime/Slider.java]: @halatmit Superfluous import...

**Comment by jisqyv** [comment]
@halatmit Hmm. When running on Gingerbread, you can enable and disable the Thumb in the Designer, but not the blocks. I'm poking at it...

**Comment by halatmit** [comment]
Hmm...  this is my mistake.  I had misread the documentation to think that
the required method is seekbar.setEnabled, but that's apparently the wrong
thing.

Jeff,

Do you want to look at this?  Or should I do that when I get back?





== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by jisqyv** [comment]
I'll look it over... It would be nice to get the fix into this upcoming
release.

-Jeff

On Fri, Aug 7, 2015 at 1:58 PM, hal <notifications@github.com> wrote:

> Hmm... this is my mistake. I had misread the documentation to think that
> the required method is seekbar.setEnabled, but that's apparently the wrong
> thing.
>
> Jeff,
>
> Do you want to look at this? Or should I do that when I get back?
>
>
>
>
>
> == Hal Abelson
> hal@mit.edu
> Prof. of Comp. Sci. and Eng.
> MIT Dept. of Elec. Eng. and Comp. Sci.
> MIT Media Lab, Second Class Faculty
>
>
> On Fri, Aug 7, 2015 at 1:39 PM, Jeffrey Schiller <notifications@github.com
> >
> wrote:
>
> > @halatmit <https://github.com/halatmit> Hmm. When running on
> Gingerbread,
> > you can enable and disable the Thumb in the Designer, but not the blocks.
> > I'm poking at it...
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/mit-cml/appinventor-sources/pull/478#issuecomment-128774911
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-cml/appinventor-sources/pull/478#issuecomment-128779004>
> .
>



-- 
_______________________________________________________________________
Jeffrey I. Schiller
Information Systems and Technology – MIT App Inventor
Massachusetts Institute of Technology
77 Massachusetts Avenue  Room E17-110A, 32-392
Cambridge, MA 02139-4307
617.910.0259 - Voice
jis@mit.edu
http://jis.qyv.name
http://appinventor.mit.edu
_______________________________________________________________________

**Comment by halatmit** [comment]
Thanks.   I think I just misread the "enabled" operation.  Maybe that was
the wrong command.



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by jisqyv** [comment]
@halatmit @afmckinney I got this one sorted. I'll make the minor change needed and push it to Gerrit.

Tested on Gingerbread and KitKat (and will test on Donut for completeness).

File: appinventor/components/src/com/google/appinventor/components/runtime/Slider.java (hunk 1 above)
Comment (jisqyv) [inline appinventor/components/src/com/google/appinventor/components/runtime/Slider.java]: @halatmit I'm going to flush the warning. It isn't needed.

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java [hunk 3]
```diff
@@ -5115,7 +5115,13 @@ String newerVersionComponentException(String componentType, int srcCompVersion,
...
   String SharingHelpStringComponentPallette();
 
-  @DefaultMessage("A Slider is a progress bar that adds a draggable thumb. You can touch the thumb and drag left or right to set the slider thumb position. As the Slider thumb is dragged, it will trigger the PositionChanged event, reporting the position of the Slider thumb. The reported position of the Slider thumb can be used to dynamically update another component attribute, such as the font size of a TextBox or the radius of a Ball.")
+  @DefaultMessage("A Slider is a progress bar that adds a draggable thumb. You can touch " +
+        "the thumb and drag left or right to set the slider thumb position. " +
+        "As the Slider thumb is dragged, it will trigger the PositionChanged event, " +
+        "reporting the position of the Slider thumb. The reported position of the " +
+        "Slider thumb can be used to dynamically update another component " +
+        "attribute, such as the font size of a TextBox or the radius of a Ball." +
+        "Sliders require Android version at least 4.1 (API Level 16).")
```
Comment (jisqyv) [inline appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java]: I'm going to eliminate the language about needing API level 16, you don't.

File: appinventor/components/src/com/google/appinventor/components/runtime/Slider.java [hunk 4]
```diff
@@ -69,6 +73,23 @@
...
+  // We do this trick by putting the use of getThumb in the class SeekBarHelper and arranging for
+  // the class to be compiled only if the API level is at least Gingerbread.  This same trick is
+  // used in implementing the Sound component.
+  private class SeekBarHelper {
+    public void getThumb(int alpha) {
+      seekbar.getThumb().mutate().setAlpha(alpha);
+    }
+  }
+
+  public final boolean referenceGetThumb = (SdkLevel.getLevel() >= SdkLevel.LEVEL_GINGERBREAD);
```
Comment (jisqyv) [inline appinventor/components/src/com/google/appinventor/components/runtime/Slider.java]: @halatmit I'm going to change LEVEL_GINGERBREAD to LEVEL_JELLYBEAN (that's the magic!).

**Comment by halatmit** [comment]
Thanks for finding that.


== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by jisqyv** [comment]
@halatmit So if you use a system < Jelly Bean, the slider thumb does not disappear when it is disabled, but it cannot be moved.

**Comment by jisqyv** [comment]
@halatmit On Gerrit for the UCR branch

**Comment by afmckinney** [comment]
Great thanks Jeff.

Andrew F. McKinney
andrew.f.mckinney@gmail.com <mailto:andrew.f.mckinney@gmail.com>
+1 978-255-3979

**Comment by jisqyv** [comment]
@halatmit @afmckinney LGTM, Merged into UCR

**Pull request closed without merging**